tunnel/tunnel.go

```diff
@@ -1120,13 +1120,18 @@ type Client struct {
 	cancelFunc func()
 }
 
-func (c *Client) cancel(err error) {
+// Cancel cancels goroutine from Start() and streamHandler(), and records error.
+func (c *Client) Cancel(err error) {
+	// Avoid calling multiple times.
+	if c.cancelFunc == nil {
```

Comment:
This should be done in the protected section, ie. after the `Lock()` call.

Reply:
Yes, `defer` is probably a better solution here.